File: Levenshtein_distance/Go/jcla1/levenshtein.go
package levenshtein

// Warning: Very slow!
func RecursiveLevenshtein(as, bs string) int {
	if min(len(as), len(bs)) == 0 {
		return len(as) | len(bs)
	}

	a, b := as[0], bs[0]
	cost := bool2Int(a == b)

	return min(1+RecursiveLevenshtein(as[1:], bs),
		1+RecursiveLevenshtein(as, bs[1:]),
		cost+RecursiveLevenshtein(as[1:], bs[1:]))
}

func WagnerFisherLevenshtein(as, bs string) int {
	m, n := len(as), len(bs)

	d := make([][]int, m+1)
	for i := range d {
		d[i] = make([]int, n+1)
	}

	for i := 0; i <= m; i++ {
		d[i][0] = i
	}

	for i := 0; i <= n; i++ {
		d[0][i] = i
	}

	for j := 1; j <= n; j++ {
		for i := 1; i <= m; i++ {
			if as[i-1] == bs[j-1] {
				d[i][j] = d[i-1][j-1]
			} else {
				d[i][j] = min(d[i-1][j]+1, d[i][j-1]+1, d[i-1][j-1]+1)
			}
		}
	}

	return d[m][n]
}

func ReducedMatrixLevenshtein(as, bs string) int {
	if min(len(as), len(bs)) == 0 {
		return len(as) | len(bs)
	} else if as == bs {
		return 0
	}

	v0 := make([]int, len(bs)+1)
	for i := range v0 {
		v0[i] = i
	}

	v1 := make([]int, len(bs)+1)

	for i := range as {
		v1[0] = i + 1

		for j := range bs {
			cost := bool2Int(as[i] == bs[j])
			v1[j+1] = min(v1[j]+1, v0[j+1]+1, v0[j]+cost)
		}

		for i := range v0 {
			v0[i] = v1[i]
		}
	}

	return v1[len(bs)]
}

func bool2Int(a bool) int {
	if a {
		return 0
	}
	return 1
}

func minSingle(a, b int) int {
	if a > b {
		return b
	}
	return a
}

func min(as ...int) int {
	if len(as) == 0 {
		return 0
	} else if len(as) == 1 {
		return as[0]
	}

	m := as[0]
	for i := 1; i < len(as); i++ {
		m = minSingle(m, as[i])
	}

	return m
}

func max(a, b int) int {
	if a < b {
		return b
	}
	return a
}

File: Levenshtein_distance/Javascript/jcla1/levenshtein.js
function recursiveLevenshtein(a, b) {
    if (min(a.length, b.length) == 0) return a.length | b.length;

    var cost = (a[0] != b[0]) * 1;

    return min(
        1 + recursiveLevenshtein(a.slice(1), b),
        1 + recursiveLevenshtein(a, b.slice(1)),
        cost + recursiveLevenshtein(a.slice(1), b.slice(1))
        );
}

function wagnerFisherLevenshtein(a, b) {
    var m = a.length,
        n = b.length,
        d = range(m + 1, range(n + 1, 0));

    for (var i = 0; i <= n; i++) {
        d[0][i] = i;
    }

    for (var j = 1; j <= n; j++) {
        for (var i = 1; i <= m; i++) {
            if (a[i-1] == b[j-1]) {
                d[i][j] = d[i-1][j-1];
            } else {
                d[i][j] = min(d[i-1][j]+1, d[i][j-1]+1, d[i-1][j-1]+1);
            }
        }
    }

    return d[m][n];
}

function reducedMatrixLevenshtein(a, b) {
    if (min(a.length, b.length) == 0) {
        return a.length | b.length;
    } else if (a === b) {
        return 0;
    }

    var v0 = range(b.length + 1),
        v1 = range(b.length + 1, 0);

    for (var i = 0; i < a.length; i++) {
        v1[0] = i + 1;

        for (var j = 0; j < b.length; j++) {
            var cost = (a[i] != b[j]) * 1;
            v1[j+1] = min(v1[j]+1, v0[j+1]+1, v0[j]+cost);
        }

        for (var k = 0; k < v0.length; k++) {
            v0[k] = v1[k];
        }
    }

    return v1[b.length];
}

function range(n, elem) {
    return Array.apply(null, Array(n)).map(function(_, i) {
        if (elem === undefined) return i;
        return elem;
    });
}

function min() {
    var arr = Array.prototype.slice.call(arguments, 0)
    return arr.reduce(function(a, b) {
        if (a > b) return b;
        return a;
    });
}